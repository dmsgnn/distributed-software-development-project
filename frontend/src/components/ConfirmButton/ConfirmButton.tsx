import { Box } from '@chakra-ui/layout';
import {
	Button,
	Modal,
	ModalBody,
	ModalCloseButton,
	ModalContent,
	ModalFooter,
	ModalHeader,
	ModalOverlay,
	useDisclosure,
} from '@chakra-ui/react';
import { useRouter } from 'next/router';
import React from 'react';

interface ConfirmButtonProps {
	onSuccessAction: () => void;
	headerText: string;
	bodyText: string;
	buttonText: string;
	isDanger?: boolean;
	redirectLink: string;
}

export const ConfirmButton = ({
	onSuccessAction,
	buttonText,
	headerText,
	bodyText,
	isDanger,
	redirectLink,
}: ConfirmButtonProps) => {
	const { isOpen, onOpen, onClose } = useDisclosure();
	const router = useRouter();

	const onSubmit = () => {
		onSuccessAction();
		onClose();
		router.push(redirectLink);
	};

	return (
		<>
			<Button onClick={onOpen} colorScheme={isDanger ? 'red' : ''}>
				{buttonText}
			</Button>

			<Modal isOpen={isOpen} onClose={onClose}>
				<ModalOverlay />
				<ModalContent>
					<ModalHeader>{headerText}</ModalHeader>
					<ModalCloseButton />
					<ModalBody>
						<Box>{bodyText}</Box>
					</ModalBody>

					<ModalFooter>
						<Button variant="ghost" onClick={onClose} mr={3}>
							Close
						</Button>
						<Button colorScheme={isDanger ? 'red' : ''} onClick={onSubmit}>
							{buttonText}
						</Button>
					</ModalFooter>
				</ModalContent>
			</Modal>
		</>
	);
};
