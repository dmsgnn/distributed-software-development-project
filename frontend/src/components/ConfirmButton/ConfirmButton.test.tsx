import { axe } from 'jest-axe';

import { act, render, screen, waitForElementToBeRemoved } from 'test-utils';
import { ConfirmButton } from './ConfirmButton';

const mockRouterPush = jest.fn();
jest.mock('next/router', () => ({ useRouter: jest.fn(() => ({ push: mockRouterPush })) }));

describe('ConfirmButton', () => {
	const props: Parameters<typeof ConfirmButton>[0] = {
		bodyText: 'bodyText',
		buttonText: 'buttonText',
		headerText: 'headerText',
		isDanger: false,
		onSuccessAction: jest.fn(),
		redirectLink: 'redirectLink',
		closeButtonText: 'closeButtonText',
	};

	it('should matches snapshot', () => {
		const { asFragment } = render(<ConfirmButton {...props} />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should be accessible', async () => {
		const { container } = render(<ConfirmButton {...props} />);

		let results: Awaited<ReturnType<typeof axe>>;
		await act(async () => {
			results = await axe(container);
		});

		expect(results).toHaveNoViolations();
	});

	it('should open modal on button click ', async () => {
		render(<ConfirmButton {...props} />);

		const button = screen.getByRole('button', { name: props.buttonText });
		expect(button).toBeInTheDocument();

		expect(screen.queryByRole('dialog')).toBeFalsy();
		await act(async () => {
			button.click();
		});

		expect(await screen.findByRole('dialog')).toBeTruthy();
	});

	it("should close modal on 'Close' button click", async () => {
		render(<ConfirmButton {...props} />);

		const button = screen.getByRole('button', { name: props.buttonText });
		expect(button).toBeInTheDocument();

		await act(async () => {
			button.click();
		});

		const closeButton = screen.getByRole('button', { name: props.closeButtonText });
		expect(closeButton).toBeInTheDocument();

		await act(async () => {
			closeButton.click();
		});

		await waitForElementToBeRemoved(screen.queryByRole('dialog'));
		expect(screen.queryByRole('dialog')).toBeFalsy();
	});

	it('should call onSuccessAction and redirect on submit', async () => {
		render(<ConfirmButton {...props} />);

		const button = screen.getByRole('button', { name: props.buttonText });
		expect(button).toBeInTheDocument();

		await act(async () => {
			button.click();
		});

		const submitButton = screen.getByRole('button', { name: props.buttonText });
		expect(submitButton).toBeInTheDocument();

		await act(async () => {
			submitButton.click();
		});

		expect(props.onSuccessAction).toBeCalledTimes(1);
		expect(mockRouterPush).toBeCalledTimes(1);
		expect(mockRouterPush).toBeCalledWith(props.redirectLink);
	});
});
