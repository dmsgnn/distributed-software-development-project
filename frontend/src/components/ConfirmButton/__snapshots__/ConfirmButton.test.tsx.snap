// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmButton should matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-4xx2wk"
    type="button"
  >
    buttonText
  </button>
</DocumentFragment>
`;
